File: src/db/ccu_store.cpp
#include "db/ccu_store.hpp"
#include <sstream>

CCUStore::CCUStore(std::string db_name)
    : db_instance{}
{
    this->db_name = db_name;
}

/**
 * Saves the given task to a database as a new document under the "tasks" collection
 *
 * @param task a reference to a const ccu::Task object representing a task
 */
void CCUStore::addTask(const ccu::Task& task)
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto collection = db_client[this->db_name]["tasks"];
    bsoncxx::builder::stream::document document{};
    //TODO: save the current timestamp
    Json::Value task_json = task.toJson();
    std::string task_string = Json::writeString(this->json_stream_builder, task_json);
    bsoncxx::document::value value = bsoncxx::from_json(task_string);
    collection.insert_one(value.view());
}

/**
 * Saves the given task to a database as a new document under the "task_archive" collection
 * and deletes it from the "tasks", "ongoing_tasks", and "ongoing_task_status" collections
 *
 * @param task a previously scheduled task
 * @param task_status status task status description
 */
void CCUStore::archiveTask(ccu::Task task, ccu::TaskStatus task_status)
{
    mongocxx::client db_client{mongocxx::uri{}};

    //adding the task to the "task_archive" collection
    auto archive_collection = db_client[this->db_name]["task_archive"];
    bsoncxx::builder::stream::document document{};
    Json::Value task_json = task.toJson();
    //TODO: save the current timestamp
    task_json["task_status"] = task_status.status;
    for (auto robot_actions : task.robot_actions)
    {
        std::string robot_id = robot_actions.first;
        std::vector<std::string> completed_actions = task_status.completed_robot_actions[robot_id];

        Json::Value &robot_action_list = task_json["robot_actions"][robot_id];
        for (auto robot_action : robot_action_list)
        {
            std::string action_id = robot_action["id"].asString();
            if (std::find(completed_actions.begin(),
                          completed_actions.end(), action_id)
                          != completed_actions.end())
            {
                robot_action["status"] = "completed";
            }
        }
    }

    std::string task_string = Json::writeString(this->json_stream_builder, task_json);
    bsoncxx::document::value value = bsoncxx::from_json(task_string);
    archive_collection.insert_one(value.view());

    //removing the task from the "ongoing_tasks" collection
    auto ongoing_task_collection = db_client[this->db_name]["ongoing_tasks"];
    ongoing_task_collection.delete_one(bsoncxx::builder::stream::document{}
                                       << "task_id" << task.id
                                       << bsoncxx::builder::stream::finalize);

    //removing the task from the "ongoing_task_status" collection
    auto task_status_collection = db_client[this->db_name]["ongoing_task_status"];
    task_status_collection.delete_one(bsoncxx::builder::stream::document{}
                                      << "task_id" << task.id
                                      << bsoncxx::builder::stream::finalize);

    //removing the task from the "tasks" collection
    auto scheduled_task_collection = db_client[this->db_name]["tasks"];
    scheduled_task_collection.delete_one(bsoncxx::builder::stream::document{}
                                         << "id" << task.id
                                         << bsoncxx::builder::stream::finalize);
}

/**
 * Saves the given task id to a database as a new document under the "ongoing_tasks" collection
 *
 * @param task_id UUID representing the id of an already scheduled task
 */
void CCUStore::addOngoingTask(std::string task_id)
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto collection = db_client[this->db_name]["ongoing_tasks"];
    bsoncxx::builder::stream::document document{};
    //TODO: save the current timestamp
    document << "task_id" << task_id;
    collection.insert_one(document.view());
}

/**
 * Adds a new task status document under the "ongoing_task_status" collection
 *
 * @param task_status task status description
 */
void CCUStore::addTaskStatus(const ccu::TaskStatus& task_status)
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto collection = db_client[this->db_name]["ongoing_task_status"];
    bsoncxx::builder::stream::document document{};
    //TODO: save the current timestamp
    Json::Value task_status_json = task_status.toJson();
    std::string task_status_string = Json::writeString(this->json_stream_builder, task_status_json);
    bsoncxx::document::value value = bsoncxx::from_json(task_status_string);
    collection.insert_one(value.view());
}

/**
 * Saves an updated status for the given task under the "ongoing_task_status" collection
 *
 * @param task_status task status description
 */
void CCUStore::updateTaskStatus(const ccu::TaskStatus& task_status)
{

}

/**
 * Returns a vector of ids representing all tasks that are saved
 * under the "ongoing_tasks" collection
 */
std::vector<std::string> CCUStore::getOngoingTasks()
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto database = db_client[this->db_name];
    auto collection = database["ongoing_tasks"];
    auto cursor = collection.find({});

    std::vector<std::string> task_ids;
    for (auto doc : cursor)
    {
        std::string task_id = doc["task_id"].get_utf8().value.to_string();
        task_ids.push_back(task_id);
    }
    return task_ids;
}

/**
 * Returns a dictionary of task IDs and ccu::Task objects representing
 * the scheduled tasks that are saved under the "tasks" collection
 */
std::map<std::string, ccu::Task> CCUStore::getScheduledTasks()
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto database = db_client[this->db_name];
    auto collection = database["tasks"];
    auto cursor = collection.find({});

    std::map<std::string, ccu::Task> scheduled_tasks;
    for (auto doc : cursor)
    {
        std::string task_id = doc["id"].get_utf8().value.to_string();
        ccu::Task task = this->getTask(task_id);
        scheduled_tasks[task_id] = task;
    }
    return scheduled_tasks;
}

/**
 * Returns a ccu::Task object representing the task with the given id
 *
 * @param task_id UUID representing the id of a task
 */
ccu::Task CCUStore::getTask(std::string task_id)
{
    mongocxx::client db_client{mongocxx::uri{}};
    auto database = db_client[this->db_name];
    auto collection = database["tasks"];
    auto doc = collection.find_one(bsoncxx::builder::stream::document{}
                                   << "id" << task_id
                                   << bsoncxx::builder::stream::finalize);
    std::string json_doc = bsoncxx::to_json((*doc));
    ccu::Task task = ccu::Task::fromJson(json_doc);
    return task;
}

File: fleet_management/db/ccu_store.py
import logging
from datetime import timezone, datetime

import pymongo as pm
from pymongo.errors import ServerSelectionTimeoutError
from ropod.structs.area import SubArea, SubAreaReservation
from ropod.structs.status import TaskStatus
from mrs.structs.timetable import Timetable


class CCUStore(object):
    """An interface for saving CCU data into and retrieving them from a database

    @author Alex Mitrevski, Argentina Ortega Sainz
    @contact aleksandar.mitrevski@h-brs.de, argentina.ortega@h-brs.de
    """

    def __init__(self, db_name='ccu_store', port=27017, **_):
        self.logger = logging.getLogger('fms.db')
        self.db_name = db_name
        self.db_port = port

        try:
            # Default timeout is 30s
            self.client = pm.MongoClient(port=self.db_port)
            self.logger.debug(self.client.server_info())
        except ServerSelectionTimeoutError as err:
            self.logger.critical("Cannot connect to MongoDB", exc_info=True)
            return

        self.db = self.client[self.db_name]
        self.logger.info("Connected to %s on port %s", self.db_name, self.db_port)

    def __str__(self):
        return str(self.__dict__)

    def unique_insert(self, collection, dict_to_insert, key, value):
        """Inserts an element to a given collection but only if it's key doesn't
           already exist.
        """
        found_dict = collection.find_one({key: value})

        if found_dict is None:
            collection.insert(dict_to_insert)
        else:
            self.logger.warning("Element: %s already exist. Not adding!", dict_to_insert)

    def archive_task(self, task, task_status):
        """Saves the given task to a database as a new document under the "task_archive" collection.

        Keyword arguments:
        @param task a previously scheduled task
        @param task_status task status description
        """

        # adding the task to the "task_archive" collection
        dict_task = task.to_dict()
        # TODO: save the current timestamp
        dict_task['task_status'] = task_status.status
        for robot_id in task.robot_actions:
            completed_actions = task_status.completed_robot_actions[robot_id]
            for i in range(len(task.robot_actions[robot_id])):
                actions = task.robot_actions[robot_id]
                for action in actions:
                    if action.id in completed_actions:
                        task.robot_actions[robot_id]['status'][i] = 'completed'

        archive_collection = self.db['task_archive']
        archive_collection.insert_one(dict_task)

        # removing the task from the "ongoing_tasks" collection
        ongoing_tasks_collection = self.db['ongoing_tasks']
        ongoing_tasks_collection.delete_one({'task_id': task.id})

        # removing the task from the "ongoing_task_status" collection
        task_status_collection = self.db['ongoing_task_status']
        task_status_collection.delete_one({'task_id': task.id})

        # removing the task from the "tasks" collection
        scheduled_task_collection = self.db['tasks']
        scheduled_task_collection.delete_one({'id': task.id})

    def add_sub_area(self, sub_area):
        """Adds sub area to the sub_areas table.

        Keyword arguments:
          @param sub_area sub area object
        """
        collection = self.db['sub_areas']
        dict_sub_area = sub_area.to_dict()
        self.unique_insert(collection, dict_sub_area, 'id', dict_sub_area['id'])

    def get_sub_area(self, sub_area_id):
        """Get sub area from sub_areas table using id.

        Keyword arguments:
        @param sub_area_id sub area id (int)
        """
        collection = self.db['sub_areas']
        sub_area_dict = collection.find_one({'id': sub_area_id})
        return SubArea.from_dict(sub_area_dict)

    def get_sub_areas(self, type):
        """Get sub areas based on type.

        Keyword arguments:
        @param type  sub area type (string)
        """
        collection = self.db['sub_areas']
        sub_area_dicts = collection.find({'type': type})
        sub_areas = []
        for sub_area_dict in sub_area_dicts:
            sub_areas.append(SubArea.from_dict(sub_area_dict))
        return sub_areas

    def delete_sub_areas(self):
        """Deletes all sub areas (used only for unit testing).
        """
        collection = self.db['sub_areas']
        res = collection.delete_many({})
        return res.deleted_count

    def add_sub_area_reservation(self, sub_area_reservation):
        """Adds new sub area reservation.

        Keyword arguments:
        @param sub_area_reservation object
        """
        collection = self.db['sub_areas_reservations']
        dict_sub_area_reservation = sub_area_reservation.to_dict()
        return collection.insert_one(dict_sub_area_reservation).inserted_id

    def get_sub_area_reservation(self, sub_area_reservation_id):
        """Gets sub area reservations.

        Keyword arguments:
        @param sub_area_reservation_id (int)
        """
        collection = self.db['sub_areas_reservations']
        sub_area_reservation_dict = collection.find_one({'_id': sub_area_reservation_id})
        return SubAreaReservation.from_dict(sub_area_reservation_dict)

    def delete_sub_area_reservations(self):
        """Deletes all sub area reservations (used only for unit testing).
        """
        collection = self.db['sub_areas_reservations']
        res = collection.delete_many({})
        return  res.deleted_count

    def get_all_future_reservations(self, sub_area_id):
        """Gets all future reservations for given sub area.

        Keyword arguments:
        @param sub_area_id (int)
        """
        collection = self.db['sub_areas_reservations']
        # TODO This doesn't match the Unix time timestamp used in the rest of the code base
        future_reservation_dict_list = collection.find({'subAreaId': sub_area_id,
                                                        'startTime': {'$gte': datetime.now(timezone.utc).isoformat()}})
        future_reservations = []
        for future_reservation_dict in future_reservation_dict_list:
            future_reservations.append(SubAreaReservation.from_dict(future_reservation_dict))
        return future_reservations

    def update_sub_area_reservation(self, sub_area_reservation_id, status):
        """Updates of existing sub area reservation.

        Keyword arguments:
        @param sub_area_reservation_id (int)
        @param status "unknown or scheduled or cancelled" (string)
        """
        collection = self.db['sub_areas_reservations']

        sub_area_reservation = self.get_sub_area_reservation(sub_area_reservation_id)
        sub_area_reservation.status = status
        dict_sub_area_reservation = sub_area_reservation.to_dict()
        return collection.replace_one({'_id': sub_area_reservation_id}, dict_sub_area_reservation)

    def clean(self):
        self.client.drop_database(self.db_name)
